receiver/prometheusreceiver/internal/transaction.go

```diff
@@ -197,15 +216,64 @@ func (t *transaction) AppendExemplar(_ storage.SeriesRef, l labels.Labels, e exe
 		return 0, errMetricNameNotFound
 	}
 
-	mf := t.getOrCreateMetricFamily(getScopeID(l), mn)
+	mf, _ := t.getOrCreateMetricFamily(getScopeID(l), mn)
 	mf.addExemplar(t.getSeriesRef(l, mf.mtype), e)
 
 	return 0, nil
 }
 
-func (t *transaction) AppendHistogram(_ storage.SeriesRef, _ labels.Labels, _ int64, _ *histogram.Histogram, _ *histogram.FloatHistogram) (storage.SeriesRef, error) {
-	//TODO: implement this func
-	return 0, nil
+func (t *transaction) AppendHistogram(_ storage.SeriesRef, ls labels.Labels, atMs int64, h *histogram.Histogram, fh *histogram.FloatHistogram) (storage.SeriesRef, error) {
+	select {
+	case <-t.ctx.Done():
+		return 0, errTransactionAborted
+	default:
+	}
+
+	if len(t.externalLabels) != 0 {
+		ls = append(ls, t.externalLabels...)
+		sort.Sort(ls)
+	}
+
+	if t.isNew {
+		if err := t.initTransaction(ls); err != nil {
+			return 0, err
+		}
+	}
+
+	// Any datapoint with duplicate labels MUST be rejected per:
+	// * https://github.com/open-telemetry/wg-prometheus/issues/44
+	// * https://github.com/open-telemetry/opentelemetry-collector/issues/3407
+	// as Prometheus rejects such too as of version 2.16.0, released on 2020-02-13.
+	if dupLabel, hasDup := ls.HasDuplicateLabelNames(); hasDup {
+		return 0, fmt.Errorf("invalid sample: non-unique label names: %q", dupLabel)
+	}
+
+	metricName := ls.Get(model.MetricNameLabel)
+	if metricName == "" {
+		return 0, errMetricNameNotFound
+	}
+
+	// The `up`, `target_info`, `otel_scope_info` metrics should never generate native histograms,
+	// thus we don't check for them here as opposed to the Append function.
+
+	curMF, existing := t.getOrCreateMetricFamily(getScopeID(ls), metricName)
+	if !existing {
+		curMF.mtype = pmetric.MetricTypeExponentialHistogram
+	} else if curMF.mtype != pmetric.MetricTypeExponentialHistogram {
+		// Already scraped as classic histogram.
```

Comment:
is the classic histogram append guaranteed to come first?

Reply:
It's the other way around. No guarantee beyond that this is how it's [coded](https://github.com/prometheus/prometheus/blob/b37258c99b901490ccf22f85f472ba2ba8c8be33/model/textparse/protobufparse.go#L72) right now, which is why I added the test case `TestNativeVsClassicHistogramScrapeViaProtobuf` .